src/permutations.jl

```diff
@@ -9,14 +9,59 @@ export
     permutations
 
 
-struct Permutations{T}
-    a::T
-    t::Int
+struct PermutationIterator{T}
+    data::T
+    length::Int
+end
+
+function has_repeats(state::Vector{Int})
+    for outer in eachindex(state)
+        for inner in (outer+1):lastindex(state)
+            if state[outer] == state[inner]
+                return true
+            end
+        end
+    end
+    return false
 end
 
-Base.eltype(::Type{Permutations{T}}) where {T} = Vector{eltype(T)}
+function increment!(state::Vector{Int}, min::Int, max::Int)
+    state[end] += 1
+    for i in reverse(eachindex(state))[firstindex(state):end-1]
+        if state[i] > max
+            state[i] = min
+            state[i-1] += 1
+        end
+    end
+end
+
+function next_permutation!(state::Vector{Int}, min::Int, max::Int)
+    while true
+        increment!(state, min, max)
+        has_repeats(state) || break
+    end
+end
+
+function Base.iterate(p::PermutationIterator, state::Vector{Int}=Int[])
+    if isempty(state)
+        # case for invalid t; iterator is empty
+        p.length > length(p.data) && return nothing
+        state = fill(firstindex(p.data), p.length)
+    end
+    # this is a special case if t=0 we return state [1] below which allows us to terminate early here
+    length(state) == 1 && p.length == 0 && return nothing
+    if !isempty(state)
+        next_permutation!(state, firstindex(p.data), lastindex(p.data))
+        if state[firstindex(state)] > lastindex(p.data)
+            return nothing
+        end
+    end
+    return [p.data[i] for i in state], p.length == 0 ? [1] : state
+end
+
+Base.length(p::PermutationIterator) = Int(div(factorial(big(length(p.data))), factorial(big(length(p.data) - p.length))))
```

Comment:
```suggestion
function Base.length(p::PermutationIterator)
    length(p.data) < p.length && return 0
    return Int(prod(big(length(p.data) - p.length + 1):big(length(p.data))
end
```
Also if you define `Base.length` and `Base.eltype` they should be tested for all cases in tests.

Reply:
added